### ISSUE TITLE
Bracket matcher/finder and auto-completion

### ISSUE THREAD

**Issue by LukeHSC** [rubberduck-vba/Rubberduck#4033]
hi,  
a bracket matcher/ finder and auto-completion would be great. 

**bracket matcher** 

- For example, within a Excel equation within a cell the Excel shows you different colours so you can see where the matching bracket in your equation or function is. This would be tremendously useful In RubberDuck IDE. - 
- For example in UltraEdit you can press Ctrl+B and it will find the matching bracket for you:
![image](https://user-images.githubusercontent.com/39577280/40462756-b7b1648a-5f54-11e8-8a00-5201aa62b29e.png)

**auto-completion**

- an autocomplete , as you type, for things like brackets , for and do loops ...

**Comment by retailcoder** [comment]
💯 Agreed, would be awesome.

However the VBIDE API exposes exactly zero code pane events, so autocompletion needs to be implemented in (or more precisely *invoked by*) the `VBENativeServices.VbeEventCallback` method.

This has been on the roadmap since forever 😄 

We need to think of a neat API that makes it easy to use the same "hook" to later implement similar/related features, i.e. "insert [this code] when [this token] is typed in the editor". I can see a matching closing brace automatically inserted relatively easily - whether it's a `(`, a `{`, or a `[`.

Highlighting brackets, or anything involving how code is rendered in the VBE, is out of the question though... for now. Ultimately we will be replacing the *entire* code pane client area with our own editor, and then sky's pretty much the limit... as long as the VBE isn't in break mode, but that's another discussion!

I think we can implement the hooks without having to wait for our own code panes.

Thanks for the feedback!

**Comment by LukeHSC** [comment]
that would be great, at least find matching brace … even better find matching IF / ENDIF  and FOR/ENDFOR !

**Comment by retailcoder** [comment]
So, uh, I totally got nerd-sniped, and got something like a skeleton of an autocompletion feature. needs refactoring and polishing and then a settings UI to enable/disable them, but anyway at this point I have `(` autocompleting to `()`, `"` autocompleting to `""`, then `[` to `]` and `{` to `}`, and `If True Then\n` autocompleting the `End If` on the next line (unindented), so... I need some sleep :octocat:

**Comment by retailcoder** [comment]
PR under way, watch for a pre-release build when it gets merged!

**Comment by retailcoder** [comment]
@LukeHSC merged! let me know what you think! 😃

**Comment by LukeHSC** [comment]
sorry found it … by accident when I typed a ‘(‘ it auto closed it  with ‘)’. Nice!

 

As a suggestion on the RD toobar, how about having a global turn ON/OFF button for Autocomplete?. I found that it was not easy to delete  one of the ‘(‘ without it not deleting and forcing the ‘)’ again. So I left the ‘()’ and cut and pasted the code in between that. That may be a good thing long term tho.

 

 

Regards, 

 

Luke

 

From: Luke A Burlet <Luke.Burlet@HSConsultants.net.au> 
Sent: Wednesday, 6 June 2018 6:24 PM
To: 'rubberduck-vba/Rubberduck' <reply@reply.github.com>
Subject: RE: [rubberduck-vba/Rubberduck] Bracket matcher/finder and auto-completion (#4033)

 

Hi Mathieu

 

Sorry for being daft …

 

I loaded  “v2.2.0.3347-pre” which uninstalled my older version, which is fine

 

I don’t see where on the new version the ‘bracket matching ’ is ?

 

 

Regards, 

 

Luke

**Comment by retailcoder** [comment]
@LukeHSC I like the idea of a toggle on the toolbar (although, that toolbar is getting rather wide by now), however do note that backspacing/deleting is currently broken and only effectively works when the left-most character is removed - I'm giving up on trying to devise a fix using the current VBE messenging hook; I'll swap the hooking for an actual keyhook tonight, that way I can know what key was pressed, and reliably fire *when a key is actually pressed*, instead of filtering caret messages. Autocompletion will definitely work *exactly* as intended by v2.3 "green" release!

**Comment by LukeHSC** [comment]
PS: would be nice to have a ‘matching bracket ’ finder (and IF/END etc finder) !

 

From: Luke A Burlet <Luke.Burlet@HSConsultants.net.au> 
Sent: Thursday, 7 June 2018 6:58 AM
To: 'rubberduck-vba/Rubberduck' <reply@reply.github.com>
Subject: RE: [rubberduck-vba/Rubberduck] Bracket matcher/finder and auto-completion (#4033)

 

sorry found it … by accident when I typed a ‘(‘ it auto closed it  with ‘)’. Nice!

 

As a suggestion on the RD toobar, how about having a global turn ON/OFF button for Autocomplete?. I found that it was not easy to delete  one of the ‘(‘ without it not deleting and forcing the ‘)’ again. So I left the ‘()’ and cut and pasted the code in between that. That may be a good thing long term tho.

 

 

Regards, 

 

Luke

 

From: Luke A Burlet <Luke.Burlet@HSConsultants.net.au <mailto:Luke.Burlet@HSConsultants.net.au> > 
Sent: Wednesday, 6 June 2018 6:24 PM
To: 'rubberduck-vba/Rubberduck' <reply@reply.github.com>
Subject: RE: [rubberduck-vba/Rubberduck] Bracket matcher/finder and auto-completion (#4033)

 

Hi Mathieu

 

Sorry for being daft …

 

I loaded  “v2.2.0.3347-pre” which uninstalled my older version, which is fine

 

I don’t see where on the new version the ‘bracket matching ’ is ?

 

 

Regards, 

 

Luke

**Comment by retailcoder** [comment]
Would be nice indeed, but we won't have that in the VBE's native code panes, for several reasons:

 - Selection can only be in one single contiguous place
 - Using the selection to highlight brackets would interfere with typing in a highly disruptive way
 - Matching token would have to be visible on-screen, which isn't always the case esp. for blocks
 - Changing font/weight of matching token in the code pane requires Win32/COM wizardry beyond my paygrade
 - Matching would need "inline parsing", but freshly-typed code isn't tokenized and thus isn't in the parser state yet.

Once we get to hijack the code panes themselves with our own WPF/AvalonEdit editor, we'll definitely look into implementing this. I have something like ReSharper's tooling in mind.

As for parser/resolver performance, it depends on a lot of things. Feel free to open a new issue and include the complete log for the initial parse of that project (initial parse takes the longest).

**Comment by retailcoder** [comment]
@LukeHSC note, we've just identified a bug that may cause the resolver to basically fail to resolve everything defined in a referenced library - if that's the case then MANY exceptions would be thrown (and logged), and this could very well make the whole parsing/resolving much slower than normal (and inspections might be raising a bunch of false positives too, so firing many more results than they normally would -> more peformance & memory pressure)... does pre-release 2.2.3310 perform better by any chance?

**Comment by LukeHSC** [comment]
Hi Mathieu

 

I am running as of last night

 

Version 2.2.6730.21818

OS: Microsoft Windows NT 10.0.14393.0, x64

Host Product: Microsoft Office 2016 x86

Host Version: 16.0.9330.2087

Host Executable: MSACCESS.EXE

 

I am a bit confused with the GitHub version nums as I had downloaded and installed “Rubberduck.Setup.2.2.0.3347-pre.exe”

 

I would say the refresh is about the same? I just reopened my ACCDB and did a refresh on the Code Explorer, takes 1.5 mins

**Comment by retailcoder** [comment]
Yeah there's a known issue about the "about" box showing the version# from the wrong assembly, following the split of `Rubberduck.Main` and `Rubberduck.Core`; the correct/intended version is only shown on the splash form at startup.

If you're referencing libraries that you're not using, removing them could help with performance a bit. Otherwise, watching logs for exceptions can indicate problems. Note that only the initial parse should take a while, subsequent parses only process what needs to be processed, so you can make the best use of caching by refreshing whenever you've changed anything in a module - other modules wouldn't be affected. Global variables and other identifiers referenced in modules outside where they're declared also invalidate caching of all involved modules when they're changed.

**Comment by LukeHSC** [comment]
hi Mathieu

 

Been working coding this AM with the autocomplete on, for most part seems ok, few glitches , i.e. re-autocompleting brackets if I go to edit them or backspace over them.

 

Biggest issue just now, I went to copy and paste a variable after a ‘[‘ bracket and RD made this:

                

 

I sent to turn off the autocomplete for RD but it does not want to turn off? I sent to Settings, unticked the ‘Close Square Brackets’ pressed Ok. But still autocompletes square brackets. I went back to settings and it is turned back on? Same thing with the ‘(‘ ; when I got the ‘[‘ all fixed up then the ‘(‘ repeated too much and not able to turn that off in RD settings either.

 

Tried several times on the Settings and now it is saving the setup (no autocomplete for ‘(‘ or ‘[‘ while I fixed up the issue in the screenshot above)

 

 

 

Regards, 

 

Luke

**Comment by LukeHSC** [comment]
PS: 

 

I seem to have broken RD completely. Tried to go to Settings (to turn off autocomplete) but it keeps crashing each time: 





 

Tried 4 times, same. Am removing this version and going back to previos , will wait on the stable release of autocomplete

 

 

Regards, 

 

Luke

 

From: Luke A Burlet <Luke.Burlet@HSConsultants.net.au> 
Sent: Thursday, 7 June 2018 9:50 AM
To: 'rubberduck-vba/Rubberduck' <reply@reply.github.com>; 'rubberduck-vba/Rubberduck' <Rubberduck@noreply.github.com>
Cc: 'LukeHSC' <luke@hsconsultants.net.au>; 'Mention' <mention@noreply.github.com>
Subject: RE: [rubberduck-vba/Rubberduck] Bracket matcher/finder and auto-completion (#4033)

 

hi Mathieu

 

Been working coding this AM with the autocomplete on, for most part seems ok, few glitches , i.e. re-autocompleting brackets if I go to edit them or backspace over them.

 

Biggest issue just now, I went to copy and paste a variable after a ‘[‘ bracket and RD made this:

                

 

I sent to turn off the autocomplete for RD but it does not want to turn off? I sent to Settings, unticked the ‘Close Square Brackets’ pressed Ok. But still autocompletes square brackets. I went back to settings and it is turned back on? Same thing with the ‘(‘ ; when I got the ‘[‘ all fixed up then the ‘(‘ repeated too much and not able to turn that off in RD settings either.

 

Tried several times on the Settings and now it is saving the setup (no autocomplete for ‘(‘ or ‘[‘ while I fixed up the issue in the screenshot above)

 

 

 

Regards, 

 

Luke
